### ISSUE TITLE
Support custom CA for crinit

### ISSUE THREAD

**Issue by patrickshan** [kubernetes/cluster-registry#171]
/sig multicluster

Currently crinit will generate a new self-signed CA and you can not supply your custom CA. Since we have pre-generated custom CAs for our different clusters, it would be nice if crinit could support custom CA.

For now, the work around is to manually replace ca.crt inside the credentials secret object with our own CA after crinit finished and kill the pods to have it updated with our CA.

**Comment by perotinus** [comment]
\cc @font, @pmoire @madhusudancs

@patrickshan I started to think about this a bit, and I think it's a bit trickier than it might seem. Currently, `crinit` uses one CA for three different roles:
- TLS certificate verification
- cluster registry client cert verification
- forwarding request certificate validation

In practice, each of these could be its own CA, and so we may need to provide a way to specify:
- TLS cert CA + TLS cert + TLS private key
- cluster registry client CA + cluster registry cert + cluster registry key
- request header certificate authority is set on the host cluster, so we should not need to worry about it (though I'm not quite sure about this)

That's a lot of potential command line arguments and confusion. If there is defaulting done (e.g., providing `--tls-ca` auto-sets `--client-ca` to the right value) that's going to be hard to document. Also, it's not clear to me whether request header certs/CA is necessary if someone wants to use request headers with the standalone cluster registry. Perhaps the middle ground in `crinit` is to allow one custom CA that's used for everything, and provide a guide on how to set it up with different CAs for these different roles.

Thoughts?

**Comment by patrickshan** [comment]
@perotinus based on my experience with aggregated cluster registry, it seems the most important one is the ca.crt inside the credentials secret object used by clusterregistry through `--client-ca-file` parameter. That's used to verify the request from kube-apiserver certs specified through --proxy-client-cert-file/--proxy-client-key-file parameters which is usually signed by your custom CA. As long as caBundle inside cluster registry aggregated apiservice is used to sign the server.key/server.crt used by clusterregistry app, then it should work even with crinit pre-generated CA.

**Comment by font** [comment]
@patrickshan I think I can provide some more context around this.

If the cluster registry's `--client-ca-file` parameter were used to verify the request from `kube-apiserver` certs specified in the `--proxy-client-cert-file` parameter of the `kube-apiserver`, then `kube-apiserver` would require the generated CA from cluster registry to sign that certificate prior to running. This would imply a tight coupling between the `kube-apiserver` and any add-on API server. But we do not require the `kube-apiserver` to be deployed at the same time as the cluster registry or any other add-on API server. And in our testing, we always deploy the cluster registry independently of the `kube-apiserver` itself. That's especially true when you're dealing with hosted kubernetes clusters in public cloud environments.

I believe what actually happens is that the k8s aggregator certificate specified by the `--proxy-client-cert-file` flag, is signed by the CA in the `--requestheader-client-ca-file` flag, which may be a custom CA in your case. That custom CA is unbeknownst to the cluster registry deployment. So the CA, and the `requestheader-client-ca-file` cert along with all the other aggregation layer flags, is actually published in the `extension-apiserver-authentication` configmap in the `kube-system` namespace. Therefore, aggregated API servers receiving calls from the kube-aggregator need to use that CA to perform their security checks. In order to read the CA contained in the configmap in an aggregated API server deployment such as the cluster registry, there needs to be a role binding between the `extension-apiserver-authentication-reader` role and the `serviceaccount` used for the aggregated API server. And that role binding, along with other RBAC objects, are created by `crinit` as part of the cluster registry deployment.

So in summary, I think your issues may have been specifically related to the `--requestheader-allowed-names` or related flags as discussed in #172. Do you know if just fixing the relevant flag, or removing `--requestheader-allowed-names`, without doing any cluster registry custom CA injection dance fixed your problem?

**Comment by patrickshan** [comment]
@font how do you manage the client certs verification in your testing for aggregated mode especially with hosted kubernetes cluster ? Do you need all those flags set for apiserver mentioned inside [this link](https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/#enable-apiserver-flags) ?  I am curious if we can just use a similar solution.

You're right that currently we use custom CA `--requestheader-client-ca-file` to sign certs for `--proxy-client-cert-file` .  It seems clusterregistry aggregated mode needs `--requestheader-client-ca-file` which you can either specify inside your kube-apiserver config or inside clusterregiestry config.

Even after removing `--requestheader-allowed-names` or fixing other flags for kube-apiserver config, you still need to modify ca.crt in the secret object to use your custom CA or you need to update deployment to add `--requestheader-client-ca-file` config for clusterregistry and supply your custom CA there.

**Comment by font** [comment]
@patrickshan I've tested this using `minikube` (which specifies `--requestheader-allowed-names=""`), `local-up-cluster.sh` (the flags for that were pasted farther up in this thread), and GKE. For GKE, we don't manage it as it's a hosted environment without access to the `kube-apiserver`. It just works in GKE using all of the `kube-apiserver` flags mentioned in that link, except I cannot see what the `--proxy-client-*` flags look like, and using the default `crinit aggregated init ...` deployment. That's why I'm trying to understand what is different in your environment. But here's that configmap in GKE:
```sh
→ kubectl -n kube-system get configmap extension-apiserver-authentication -o yaml
apiVersion: v1
data:
  client-ca-file: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
  requestheader-allowed-names: '["aggregator"]'
  requestheader-client-ca-file: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
  requestheader-extra-headers-prefix: '["X-Remote-Extra-"]'
  requestheader-group-headers: '["X-Remote-Group"]'
  requestheader-username-headers: '["X-Remote-User"]'
kind: ConfigMap
metadata:
...

```

I guess what I'm saying is that I think changing the `kube-apiserver` config itself, if it wasn't configured with these flags and you have access to modify it, should also suffice. This is because `kube-apiserver` supplies the CAs `--client-ca-file`, `--requestheader-client-ca-file`, etc. to any add-on API server that is registered simply by having the add-on API server access the `extension-apiserver-authentication` configmap in the `kube-system` namespace as shown above. You can take a look at that configmap in your cluster to verify. The `apiserver-builder` actually captures these details in their [auth doc](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/docs/concepts/auth.md), specifically their [client-certificate-authentication](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/docs/concepts/auth.md#client-certificate-authentication) and [requestheader-authentication](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/docs/concepts/auth.md#requestheader-authentication) sections.

In any case, I think we still need to provide a custom CA option for cluster registry. A helm chart for the cluster registry would certainly help simplify this, but there are other outstanding issues we first need to resolve before that can happen.

**Comment by patrickshan** [comment]
Thanks @font . Which version of kube are you using for test ? I upgraded my cluster to 1.9.1 today and it seems that I have similar data inside `extension-apiserver-authentication` configmap as you posted and it works after I re-deploy cluster registry using crinit without updating ca.crt. Not sure if it's because of the new kube or maybe the latest night release of cluster registry. I'll do another test on kube 1.8.6 early next week.

**Comment by font** [comment]
@patrickshan I was using k8s 1.8.X. The latest nightly release of CR shouldn't have had any impact. So it's likely it is the `kube-apiserver`. What version were you previously using?

**Comment by patrickshan** [comment]
@font I was using kube 1.8.6.

**Comment by patrickshan** [comment]
@font I just confirmed it works with those kube-apiserver parameters update on kube 1.8.6 after destroying my cluster and rebuilt it from scratch.

So feel free to close this issue, or leave it open if you want to use it to track the feature of supporting custom CA for cluster registry itself.

**Comment by font** [comment]
@patrickshan That's great, glad you got it working! It sounds like you no longer need this feature. If that's the case, I think we can close it until somebody else requests it. @perotinus WDYT?

**Comment by perotinus** [comment]
@font I'm happy to close or leave open, though I think leaving it open would make it clearer that this isn't something that we've implemented yet but that we've considered.

**Comment by perotinus** [comment]
Note that this issue is no longer applicable because of the CRD migration done in #225, which removes `crinit` and the standalone API server.
